Mouse mapper and configuration mechanism

FeralChild64 opened the pull request:
# User visible changes

- MOUSECTL.COM mouse configuration tool is now available on drive Z. Rationale: sometimes DOS mouse drivers used to provide companion tools, allowing user to set various parameters, this tools is intended to serve a similar purpose.
- `[mouse]` section is now available in the configuration file.
- It is now possible to map a single physical mouse to the given interface. _Settlers_ and _Settlers II_ can now be played by two human players on a single computer, each controlling the game with own mouse.
- 
![Screenshot_MouseMapper_Help](https://user-images.githubusercontent.com/48332137/197237023-6a365c0f-e0ee-4996-8712-656d551a3267.png)
![Screenshot-Mapping](https://user-images.githubusercontent.com/48332137/194510290-d315a113-13b2-47f7-9ed6-10a64df86ed3.png)
```ini
[mouse]
#   mouse_sensitivity: Default mouse sensitivity. Works exponentially, add 10 to double the effect.
#                      Negative values reverse mouse direction, 0 disables the movement completely.
#                      The optional second parameter specifies vertical sensitivity (e.g. 30,40).
#                      Setting can be adjusted in runtime (also per mouse interface) using internal
#                      MOUSECTL.COM tool, available on drive Z:.
#     mouse_raw_input: Enable to bypass your operating system's mouse acceleration and sensitivity
#                      settings. Works in fullscreen or when the mouse is captured in window mode.
#    dos_mouse_driver: Enable built-in DOS mouse driver.
#                      Notes:
#                         Disable if you intend to use original MOUSE.COM driver in emulated DOS.
#                         When guest OS is booted, built-in driver gets disabled automatically.
# dos_mouse_immediate: Updates mouse movement counters immediately, without waiting for interrupt.
#                      May improve gameplay, especially in fast paced games (arcade, FPS, etc.) - as
#                      for some games it effectively boosts the mouse sampling rate to 1000 Hz, without
#                      increasing interrupt overhead.
#                      Might cause compatibility issues. List of known incompatible games:
#                         - Ultima Underworld: The Stygian Abyss
#                         - Ultima Underworld II: Labyrinth of Worlds
#                      Please file a bug with the project if you find another game that fails when
#                      this is enabled, we will update this list.
#     ps2_mouse_model: PS/2 AUX port mouse model:
#                         standard:       3 buttons, standard PS/2 mouse.
#                         intellimouse:   3 buttons + wheel, Microsoft IntelliMouse.
#                      Default: intellimouse
#                      Possible values: standard, intellimouse.
#     com_mouse_model: COM (serial) port default mouse model:
#                         2button:        2 buttons, Microsoft mouse.
#                         3button:        3 buttons, Logitech mouse, mostly compatible with Microsoft mouse.
#                         wheel:          3 buttons + wheel, mostly compatible with Microsoft mouse.
#                         msm:            3 buttons, Mouse Systems mouse, NOT COMPATIBLE with Microsoft mouse.
#                         2button+msm:    Automatic choice between 2button and msm.
#                         3button+msm:    Automatic choice between 3button and msm.
#                         wheel+msm:      Automatic choice between wheel and msm.
#                      Default: wheel+msm
#                      Notes:
#                         Go to [serial] section to enable/disable COM port mice.
#                      Possible values: 2button, 3button, wheel, msm, 2button+msm, 3button+msm, wheel+msm.

mouse_sensitivity   = 50
mouse_raw_input     = true
dos_mouse_driver    = true
dos_mouse_immediate = false
ps2_mouse_model     = intellimouse
com_mouse_model     = wheel+msm
```

# Internal changes
- Imported _ManyMouse_ (https://github.com/icculus/manymouse) to allow mouse mapper implementation; library is started and stopped on demand.
- Mouse emulation code is now much more object oriented.
- It was also moved to own subdirectory - as number of files has grown, and most likely this is not the end.

# Dual mouse gaming

## _The Settlers_

If you don't have a `serial1 = mouse` set in the configuration file, execute:

```
SERIAL.COM 1 MOUSE
```

Now we have to tell DOSBox which mouse should be used for which interface using `MOUSECTL.COM`. To do this interactively, (providing name in the command line is also possible, though) execute:

```
MOUSECTL.COM DOS COM1 -map
```

Until the game is started, you have to use mouse mapped to `DOS` interface to operate the menu, the other one can only be used in-game.

Secondary mouse has to be configured in-game: OPTIONS -> RIGHT SIDE. Standard IRQ for COM1 port is 4.

With default serial mouse settings (where DOSBox Staging tries to detect what protocol the application expects) the secondary mouse has to be configured as _Mouse Systems_ mouse - game's internal _Microsoft_ serial mouse driver uses non-standard port settings, which confuses the emulator autodetection code (so far I have no idea how to improve this; might be impossible). _Microsoft_ mouse setting can be used if it is explicitly configured, for example by executing `SERIAL.COM 1 MOUSE model:2button` before starting the game.

## _The Settlers II_

Use same mapping process as above - but the serial mouse cannot be a _Mouse Systems_ mouse, _Settlers II_ does not support it anymore. There is no need to configure secondary mouse in-game.

Default serial mouse settings work correctly.

Make sure the game is fully patched - _The Settlers II: Veni, Vidi, Vici_ version 1.02 does not detect secondary mouse (it is unclear if it can be worked around). _The Settlers II: Gold_ version 1.51 (GOG release) works correctly.

## Limitations

- Do not unplug the physical mouse which is mapped to emulated interface; there is no support for reconnect (not sure if it can be safely implemented with current _ManyMouse_ API), you will have to quit the game and repeat the mapping process.
- When a mouse is mapped to the emulated interface, rescanning for possible new devices is not performed - it might be possible to write a reasonably safe rescan, but with the current _ManyMouse_ API this would be tricky.
- There is no GUI for mapping the mice, but internally an API is exposed to allow writing one in the future.
- If physical mouse is mapped to the emulated interface, only first 3 mouse can be used - this might be improved in the future, but would need to be tested with different OS'es.
- Seamless mouse integration won't work correctly if the mouse is mapped; there is no way to fix this, as OS uses all the mice available in the system combined for the cursor movement

kcgen commented:
Confirmed working on Linux. 

![Screenshot from 2022-10-08 13-41-33](https://user-images.githubusercontent.com/1557255/194727531-e3089f3e-56bb-4605-9b9c-86b4aa4c5479.png)


In Settlers II, when I have the SDL window at a reasonable size (similar to other desktop windows), the mouse movement inside Staging is magnified roughly 10-fold faster that my desktop mouse moves.

As I shrink the window down, I can eventually find the window size where the mouse "inside the window" moves at the same rate as my desktop, and it's around this size:

![2022-10-08_13-55](https://user-images.githubusercontent.com/1557255/194727607-6473e126-c3a8-496a-b3a3-e8e45db01b72.png)

With the window this small, I can start with my mouse to the left of the DOSBox window, move it right into the SDL frame, and with steady motion right, the in-game cursor will touch the right side of the SDL frame just as the host cursor exits the window on the right.

But if I make the DOSBox window larger than what's show above: the mouse moves at the ratio of the larger window-to-the-smaller window, multiples faster.

kcgen commented:
Working on macOS too! IOKit is found :+1: 
Functionally behaves just like on Linux. 

Here's a video (captured on the mac) showing the cursor magnification behavior and how vertical is even a bit more magnified.

https://kcgen.duckdns.org/macos-mouse-mapper-1.mov

(On Linux, the scaling is even stronger, I would say another 2x or 4x up from this.  I recorded it on macOS because for some reason it shows the desktop cursor plus the DOS cursor at the same time.. perhaps a separate issue, but helps illustrate the scaling)

FeralChild64 commented:
@kcgen Are you trying to use seamless mouse? It won't work with mapping (I have just updated the PR description), as OS uses all the mice in the system to move the cursor, and after mapping we only use one; besides from `ManyMouse` we are only getting raw mouse events, without acceleration applied by host OS.

Kappa971 commented:
It seems that here we will need to do the same translation work as the "mixer" command.

kcgen commented:
Yup, I was using seamless (it's default). Thanks for updating the notes above.

> seamless mouse won't work with mapping

Suggest:
 - Noting in `capture_mouse`'s conf text that `seamless` mode is not compatible with multi-mouse mapping.

- Printing a DOS error message when `mousectl`'s `-map` is used with seamless.

- Logging and error on the host console (`LOG_ERR`) in this case too (for users who automate game startup in BAT files, when the console text blows by them)

![2022-10-09_08-10](https://user-images.githubusercontent.com/1557255/194764491-6888dd77-f8d7-431a-b5f8-0fbccb84cda5.png)


---

With the addition of `[mouse]` (which is great!), it would be really nice to find all the conf's mouse settings in there (including the capture state and raw toggle).

But this might involve moving the code state elements as well, as multi-mouse involves host integration, so ideally all of the mouse host and DOS states could be managed within the same logical code area.

You mentioned more refactoring/organization is likely needed, so perhaps we're thinking of the same thing :slightly_smiling_face:; 
and could be done in a follow on PR (to limit this scope).

kcgen commented:
With the addition of `[mouse]` (which is great!), it would be really nice to find all the conf's mouse settings in there (including the capture state and raw toggle).

But this might involve moving the code state elements as well, as multi-mouse involves host integration, so ideally all of the mouse host and DOS states could be managed within the same logical code area.

You mentioned more refactoring/organization is likely needed, so perhaps we're thinking of the same thing :slightly_smiling_face:; 
and could be done in a follow on PR (to limit this scope).

kcgen commented on LICENSE:
```diff
@@ -34,6 +34,7 @@ GPL-2.0-or-later terms, however there are some exceptions:
    - gui_tk: GPL-3.0-or-later
    - iir1: MIT
    - loguru: Public Domain
+   - ManyMouse: Zlib
```
Excellent addition.

kcgen commented on include/mouse.h:
```diff
@@ -22,39 +22,79 @@
+// Data types
+// ***************************************************************************
+
+enum class MouseInterfaceId : uint8_t {
+    DOS,  // emulated DOS mouse driver
+    PS2,  // PS/2 mouse (this includes VMware mouse protocol)
+    COM1, // serial mouse
+    COM2,
+    COM3,
+    COM4,
+
+    First = DOS,
+    Last  = COM4,
+    None  = UINT8_MAX
+};
+
+constexpr uint8_t num_mouse_interfaces_com = 4;
```
Specifying four COM ports here feels like an architectural island managing its own state of the world, when `include/serialport.h` is used by the serial port implementation, which has stuff like:

```c++
#define SERIAL_MAX_PORTS  4
const uint8_t serial_defaultirq[] = {4, 3, 4, 3};
const uint16_t serial_baseaddr[] = {0x3f8, 0x2f8, 0x3e8, 0x2e8};
const char *const serial_comname[] = {"COM1", "COM2", "COM3", "COM4"};
```

I faced a similar issue with the three LPT port, when refactoring the DACs, so I placed the generic LPT port info in `include/lpt.h`, which can be used by anything involving the printer ports.

Maybe there's a similar solution here, where `serial.h` and `mouse.h` both get some very basic facts about the COM ports from `comports.h`.

And then in mouse, you could have a strong programmatic link to those `ComPorts` enum class(es).

``` c++
enum class MouseInterfaceId : uint8_t {
    DOS,  // emulated DOS mouse driver
    PS2,  // PS/2 mouse (this includes VMware mouse protocol)
    COM1 = ComPorts::COM1, // serial mouse
    COM2 = ComPorts::COM2,
    COM3 = ComPorts::COM3,
    COM4 = ComPorts::COM4,
```

(Or something to that effect)

FeralChild64 replied:
Got rid of `num_mouse_interfaces_com`, reused `SERIAL_MAX_PORTS`. This is actually the only COM port related value we need to know here. Reusing port identifier from other enum class wouldn't be practical - we want to have continuous numbering (so that we can use arrays), we can't have overlapping ID's, and our number of mouse interfaces is likely to increase in the future - as InPort/Bus mouse is still not implemented.

kcgen commented on include/mouse.h:
```diff
@@ -79,4 +119,105 @@ uint8_t MOUSEBIOS_GetSampleRate();
+    MouseInterfaceId GetInterfaceId() const;
+    MouseMapStatus GetMapStatus() const;
+    const std::string &GetMappedDeviceName() const;
+    uint8_t GetSensitivityX() const; // 1-99
+    uint8_t GetSensitivityY() const; // 1-99
```
A powerful feature of DOSBox's current sensitivity is that 0 can be used to prevent input on that specific axis - for games like DOOM, so you can use "mouse-look" type actions to look left and right, and WASD keys for position sliding, with the assurance that any accident forward/back mouse movement won't "stall" your guy's movement.

Also, negative values invert the axis to help normalize those strange FPS games that unfortunately used flight-sim-type direction: "Pull back to look up", push forward to look down. :rofl: 

Adding @GranMinigun

kcgen replied:
(note for myself) reviewed up to here

kcgen commented:
Working great, @FeralChild64!

```ini
# Settlers II
# ~~~~~~~~~~~
# Multi-player with two-mouse FeralChild64's fc/mouse-mapper-1!
# 
# Pre-requisites:
#
# 1. Ensure the game version is The Settlers II: Gold version 1.51 (GOG release)
#    - The Settlers II: Veni, Vidi, Vici version 1.02 does not detect secondary mouse
#
# 2. Ensure two physical mice are plugged in and usable on the host's desktop
#
# 3. Ensure [sdl] capture_mouse = onstart (or onclick) is set
#    - Seamless mouse integration won't work correctly if the mouse is mapped
# 
# 4. Ensure [serial] serial1 = mouse is set
#    - The default serial mouse settings work correctly.
#
# Steps:
#
# 5. Launch MOUSECTL.COM DOS COM1 -map
#    - Left-click primary mouse to map it to a physical mouse
# 
# 6. Launch MOUSECTL.COM and confirm the 'DOS' mouse is not "Mouse Systems mouse"
#
[sdl]
capture_mouse = onstart
raw_mouse_input = false
sensitivity = 35,25

[cpu]
cycles=50000

[serial]
serial1 = mouse

[mouse]
mouse_dos           = true
mouse_dos_immediate = false
model_ps2           = intellimouse
model_com           = wheel+msm

[autoexec]
imgmount d "cd/The Settlers II (Gold Edition).cue" -t cdrom
mount c .
c:
mousectl dos com1 -map
mousectl
pause
cd settler2
```

![Screenshot from 2022-10-09 12-10-19](https://user-images.githubusercontent.com/1557255/194775684-d8f1c4d5-57f4-4403-b4d5-e5b4e9fd7587.png)

![Screenshot from 2022-10-09 12-11-26](https://user-images.githubusercontent.com/1557255/194775691-00298f0e-73a5-4af6-9b1d-e2a1d5f7a75f.png)

https://user-images.githubusercontent.com/1557255/194775702-47eeef2a-c2d8-46e1-a03d-4e9107560b56.mp4

kcgen commented:
Another idea to address the non-seamless compatibility issue is to programmatically capture the mouse as soon as mapping is used (and inform the user of this). 

I suspect this would be preferable for users: it's a hassle to quit out and change a conf setting (especially if the user isn't using local confs and only their primary).

FeralChild64 commented:
@kcgen: I did the following changes:

- there is a warning displayed by the _MOUSECTL_ tool if mapping is being done with potentially incompatible configuration
- mapping + seamless mouse integration => `LOG_WARNING` the first time it happens; IMHO `LOG_ERR` would be a little bit too far
- when mapping is in progress, the seamless integration is completely disabled now, DOSBox acts as if it wasn't enabled; are you sure this is a good idea to capture the mouse in such case? Someone might automate mapping in .BAT file (by supplying his physical mouse name with wildcards), sudden auto-grabbing might be confusing.

Regarding configuration:
- `capture_mouse` - moving this to `[mouse]` is indeed risky, and IMHO shouldn't be done withing this PR
- `raw_mouse_input` - this I can move to `[mouse]` section, it is not really complicated
- `sensitivity` - my proposal is to make this setting deprecated and create a new one in `[mouse]`, which would act as a default for sensitivity set by `MOUSECTL.COM` - do you agree?

Even after my changes sensitivity set by `MOUSECTL.COM` works a little bit different - it is limited to -99 - +99 range, and the effect is exponential. There are two reasons for this:
- I want to limit the amount of digits to display in the table within the DOS; we only have 80 characters in the row
- I have a future GUI in mind, and value without real upper limit, where difference between 30 and 50 is substantial but between 230 and 250 it is hardly noticeable - is rather hard to nicely reflect in the GUI with, for example, a slider

kcgen commented:
> when mapping is in progress, the seamless integration is completely disabled now, DOSBox acts as if it wasn't enabled; are you sure this is a good idea to capture the mouse in such case?

(We have a similar issue right now with `mixer.com`: if you launch with default settings, then run `mixer SB r10` and try to enable reverb just for the SB PCM output, it won't take effect because the reverb engine itself isn't running because the default `[mixer] reverb == off`; but if the user asks for it, then reverb should be activated and then just apply it to that channel)

Given mapping requires capturing the mouse, I think it perfectly reasonable to do it: and for us to also connect these concepts (capturing + mapping) in all our messaging and documentation. For example:
 - `"-map: Captures and maps a host mouse to an emulated mouse port"`
 - `"Razor1911 Mouse captured and mapped to COM2"`
 - `"Failed capturing Logitech123, PS/2 could not be mapped"`

User should be told and shown that the mouse gets captured as part of the mapping procedure.

> Someone might automate mapping in .BAT file (by supplying his physical mouse name with wildcards), sudden auto-grabbing might be confusing.

As long our wording and actions consistently explain that we capture to map, then a sophisticated user using wildcard mapping in `mousectl` will definitely understand this concept.

> Regarding configuration:
> 
>     * `capture_mouse` - moving this to `[mouse]` is indeed risky, and IMHO shouldn't be done withing this PR

:+1: ; makes sense - and might make sense to take-over and refactor the capturing/seamless state management too (pull it out of sdlmain), to fit more tightly with all your existing additions (as you see fit).

 
>     * `raw_mouse_input` - this I can move to `[mouse]` section, it is not really complicated

Right on. 

 
>     * `sensitivity` - my proposal is to make this setting deprecated and create a new one in `[mouse]`, which would act as a default for sensitivity set by `MOUSECTL.COM` - do you agree?

Yes; the `[sdl] sensitivity` vs mousectl CLI values were getting ambiguous - connecting them at the conf and CLI level makes perfect sense.

> Even after my changes sensitivity set by `MOUSECTL.COM` works a little bit different - it is limited to -99 - +99 range, and the effect is exponential. There are two reasons for this:
> 
>     * I want to limit the amount of digits to display in the table within the DOS; we only have 80 characters in the row

Yes, a range of ~200 values should have sufficient precision to span the range of scaling. 

For myself, I often want to cut sensitivity in 1/8th, 1/4th, half, or sometimes on the mac (where the mouse can move brutally slow, even on the host), I might want to 2x or 4x.

Wengier commented on src/hardware/mouse/mouse_queue.h:
```diff
@@ -0,0 +1,84 @@
+    void UpdateDelayCounters();
+    uint8_t ClampStartDelay(float value_ms) const;
+
+    struct { // intial value of delay counters, in milliseconds
```
typo: initial

FeralChild64 replied:
Fixed.

Wengier commented on include/mouse.h:
```diff
@@ -79,4 +119,107 @@ uint8_t MOUSEBIOS_GetSampleRate();
+
+    // Always destroy the object once it is not needed anymore
+    // (configuration tool finishes it's job) and we are returning
+    // to normal code excution!
```
typo: execution

FeralChild64 replied:
Fixed.

Wengier commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,537 @@
+    std::vector<std::string> params;
+    cmd->FillVector(params);
+
+    // Extract the list of interfaces from teh vector
```
Typo: the

Wengier commented on src/hardware/mouse/mouse_config.cpp:
```diff
@@ -0,0 +1,288 @@
+    const Section_prop *conf = dynamic_cast<Section_prop *>(section);
+    assert(conf);
+    if (!conf)
+        return;
```
Is this unneeded with the assert right above?

FeralChild64 replied:
Theoretically `dynamic_cast` can fail if someone attempts a configuration system rework - and this code is not being executed often. I would prefer to leave it as is.

Wengier commented on src/gui/sdlmain.cpp:
```diff
@@ -3571,35 +3580,25 @@ static void HandleMouseWheel(SDL_MouseWheelEvent *wheel)
 	
 	if (button->state == SDL_RELEASED) {
-		notify_released(button->button);
+		notify_button(button->button, false);
```
Perhaps something like this for improved readability?

`constexpr auto is_pressed = false;`
`notify_button(button->button, is_pressed);`

FeralChild64 replied:
OK, added a `constexpr`.

Wengier commented on meson.build:
```diff
@@ -614,6 +647,16 @@ if host_machine.system() == 'darwin'
+conf_data.set10('C_MANYMOUSE', true)
+if host_machine.system() == 'darwin'
+  if not conf_data.has('C_IOKIT')
+    # On OSX the ManyMouse library requires IOKit
```
Perhaps change “OSX” to read “macOS”?

FeralChild64 replied:
OK, changed.

FeralChild64 commented on src/gui/sdlmain.cpp:
```diff
@@ -3613,7 +3612,7 @@ static void HandleMouseButton(SDL_MouseButtonEvent * button)
 	}
 
-	notify_pressed(button->button);
+	notify_button(button->button, true);
```
OK, added a `constexpr`.

FeralChild64 commented:
Autocapture mouse on mapping already pushed.

kcgen commented on src/hardware/mouse/mouseif_virtual_machines.cpp:
```diff
@@ -70,21 +70,20 @@ static constexpr uint32_t ABS_UPDATED = 4;            // tells that new pointer
-static float speed = 0.0f;
+// Multiply scale by 0.02f to put acceleration_vmm in a reasonable
+// range, similar to sensitivity_dos or sensitivity_vmm)
+static MouseSpeedCalculator speed_xy(0.02f * mouse_predefined.acceleration_vmm);
```
Suggest lifting out the 0.2 scalar to a `constexpr float`.

FeralChild64 replied:
OK, changed.

kcgen commented on src/hardware/serialport/serialmouse.h:
```diff
@@ -22,81 +22,78 @@
+    // Mouse model as specified in the parameter
+    MouseModelCOM param_model = MouseModelCOM::NoMouse;
+    // If true = autoswitch between param_model and Mouse Systems mouse
+    bool param_auto_msm = false;
+
+    MouseModelCOM model = MouseModelCOM::NoMouse; // currently emulated model
+
+    uint8_t port_byte_len = 0; // how many bits the port transmits in a byte
+
+    bool has_3rd_button = false;
+    bool has_wheel      = false;
+    float rate_coeff    = 1.0f; // coefficient for boosted sampling rate
+    bool send_ack       = true;
+    uint8_t packet[6]   = {};
+    uint8_t packet_len  = 0;
+    uint8_t xmit_idx = UINT8_MAX; // index of byte to send, if >= packet_len
+                                  // it means transmission ended
+    bool xmit_2part     = false; // true = packet has a second part, which could
+                                 // not be evaluated yet
+    bool another_move   = false; // true = while transmitting a packet we
+                                 // received mouse move event
+    bool another_button = false; // true = while transmitting a packet we
+                                 // received mouse button event
```
For the bools that don't start with `has_`, suggest adding prefixing with some boolean-ness adjustectives (needs, is, has, should, etc..)

Because when I see "another_move" and "another_button" (nouns), I think they hold "move" and "button" objects. 

And when I see "xmit_2part" (verb-start), I think it's a function or a function-pointer that will transmit the "2part".

FeralChild64 replied:
Changed all except `send_ack`, as I'm not 100% sure what is the purpose of this variable.

kcgen reviewed (commented):
First round review done. This is a ton of very readable code, nicely organized into classes, enumes, and lists/vectors, and also well organized into the filesystem structure. Top notch, @FeralChild64.

Just a handful or (hopefully) useful comments that can help refine it a bit more.

FeralChild64 commented on src/dos/program_mousectl.h:
```diff
@@ -0,0 +1,76 @@
+    bool CmdSensitivityY(const std::vector<MouseInterfaceId> &list_ids);
+    bool CmdMinRate(const std::vector<MouseInterfaceId> &list_ids,
+                    const std::string &param);
+    bool CmdMinRate(const std::vector<MouseInterfaceId> &list_ids);
```
OK, cherry-picked.

FeralChild64 commented on meson.build:
```diff
@@ -568,6 +576,31 @@ if host_machine.system() == 'darwin'
+    if iokit_dep.found()
+        if cxx.check_header('IOKit/IOKitLib.h')
+            is_iokit_compilable = cxx.links(
+                iokit_code,
```
OK, changed

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,549 @@
+
+        // First parameter is consummed, remove it
+        params.erase(params.begin());
+    }
```
Thanks, cherry-picked.

FeralChild64 commented on src/hardware/mouse/mouse.cpp:
```diff
@@ -0,0 +1,673 @@
+// Interrupt 74 implementation
+// ***************************************************************************
+
+static Bitu int74_exit()
+{
+    SegSet16(cs, RealSeg(CALLBACK_RealPointer(int74_ret_callback)));
+    reg_ip = RealOff(CALLBACK_RealPointer(int74_ret_callback));
+
+    return CBRET_NONE;
+}
```
Right, these are for callback APIs. Not much can be changed here.

FeralChild64 commented on src/hardware/mouse/mouse_common.cpp:
```diff
@@ -0,0 +1,204 @@
+/*
+ *  Copyright (C) 2022       The DOSBox Staging Team
```
OK - changed everywhere in mouse code.

kcgen replied:
Thanks. The `./scripts/update-copyright-date.sh` will scan the source tree  and update the last-revised year to match the files' most recent commit year (just incase we missed something).

FeralChild64 commented on src/hardware/mouse/mouse.cpp:
```diff
@@ -0,0 +1,673 @@
+
+void MOUSE_NotifyFakePS2()
+{
+    auto interface = MouseInterface::GetPS2();
```
OK, changed.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,549 @@
+        auto compare = [&params](const char *string)
+        {
+            return (0 == strcasecmp(params[0].c_str(), string));
+        };
```
OK, changed. BTW, `strcasecmp` is also a DOSBox Staging support function; shouldn't it be made obsolete and replaced with `iequals` everywhere?

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,549 @@
+        return false;
+    }
+
+    return true;
```
OK, changed.

FeralChild64 commented on src/hardware/mouse/mouse_config.cpp:
```diff
@@ -0,0 +1,288 @@
+    "wheel+msm",
+};
+
+static std::vector<uint16_t> list_rates = {
```
OK, changes.

FeralChild64 commented on src/hardware/mouse/mouse_common.cpp:
```diff
@@ -0,0 +1,204 @@
+                                static_cast<int32_t>(INT16_MIN),
+                                static_cast<int32_t>(INT16_MAX));
+
+    return static_cast<int16_t>(tmp);
```
OK, changed.

FeralChild64 commented on src/hardware/mouse/mouse.cpp:
```diff
@@ -0,0 +1,673 @@
+// ***************************************************************************
+
+void get_relevant_interfaces(std::vector<MouseInterface *> &list_out,
+                             const std::vector<MouseInterfaceId> &list_ids)
```
OK, changed. But usually we can't use `const`.

FeralChild64 commented on src/hardware/mouse/mouse_manymouse.cpp:
```diff
@@ -0,0 +1,476 @@
+        // Do not accept already mapped devices
+        bool already_mapped = false;
+        for (auto &interface : mouse_interfaces)
+            if (interface->IsMapped(device_id))
```
OK, changed.

FeralChild64 commented on src/hardware/mouse/mouse_manymouse.cpp:
```diff
@@ -0,0 +1,476 @@
+    }
+
+    return max_mice;
+}
```
OK, added a comment.

FeralChild64 commented on src/hardware/mouse/mouse_common.cpp:
```diff
@@ -0,0 +1,204 @@
+    // This routine provides a function for mouse ballistics
+    // (cursor acceleration), to be reused by various mouse interfaces.
+    // Since this is a DOS emulator, the acceleration model is
+    // based on a historic PS/2 mouse specification.
```
Added a link.

FeralChild64 commented on src/hardware/mouse/mouse_common.cpp:
```diff
@@ -0,0 +1,204 @@
+uint16_t MOUSE_ClampRateHz(const uint16_t rate_hz)
+{
+    constexpr auto rate_min = static_cast<uint16_t>(10);
+    constexpr auto rate_max = static_cast<uint16_t>(500);
```
OK, changed.

FeralChild64 commented on src/hardware/mouse/mouse_manymouse.cpp:
```diff
@@ -0,0 +1,476 @@
+            if (name[pos - 1] == 0x7f)
+                name[pos - 1] = 0x20;
+            // Remove non-ASCII and control characters
+            if (name[pos - 1] < 0x20 || name[pos - 1] > 0x7e)
```
OK, added constants.

FeralChild64 commented on include/mouse.h:
```diff
@@ -79,4 +119,107 @@ uint8_t MOUSEBIOS_GetSampleRate();
+    // (configuration tool finishes it's job) and we are returning
+    // to normal code execution!
+
+    MouseConfigAPI();
```
This would make sense for the current usage (MOUSECTL.COM command), but the class is intended to be used also by the yet-to-be-written GUI. And here it is perfectly possible that user will want to do a different change to DOS mouse driver, and different for PS/2 mouse emulation - thus, at the moment of creating the object, we will have no knowledge regarding what content the list should have.

I propose to leave this unchanged.

kcgen replied:
Sounds good, and thanks for the explanation. I had (incorrectly) thought that the `list_ids` was _all_ of the interfaces available.

> here it is perfectly possible that user will want to do a different change to DOS mouse driver, and different for PS/2 mouse emulation

I guess `list_ids` differs based on driver - so the DOS mouse driver generates a different set of `list_ids` versus the PS/2 emulation?

FeralChild64 commented on src/hardware/mouse/mouse.cpp:
```diff
@@ -0,0 +1,673 @@
+{
+    manymouse.StopConfigAPI();
+    MOUSE_NotifyStateChanged();
+}
```
Changed name to MouseControlAPI - this is, after all, being used by MOUSE**CTL**.COM command.

kcgen replied:
> Changed name to MouseControlAPI - this is, after all, being used by MOUSE**CTL**.COM command.

Excellent, this name change alone makes is very clear how they relate.

FeralChild64 commented on src/dos/program_mousectl.h:
```diff
@@ -0,0 +1,76 @@
+    bool CmdSensitivityY(const std::vector<MouseInterfaceId> &list_ids);
+    bool CmdMinRate(const std::vector<MouseInterfaceId> &list_ids,
+                    const std::string &param);
+    bool CmdMinRate(const std::vector<MouseInterfaceId> &list_ids);
```
Added a comment with explanation.

kcgen commented on src/dos/program_mousectl.h:
```diff
@@ -0,0 +1,76 @@
+
+#include "programs.h"
+
+#include "mouse.h"
+
+
+class MOUSECTL final : public Program {
+public:
+    MOUSECTL()
+    {
+        AddMessages();
+        help_detail = {HELP_Filter::All,
+                       HELP_Category::Dosbox,
+                       HELP_CmdType::Program,
+                       "MOUSECTL"};
+    }
+    void Run();
+
```
The `./scripts/format-commit.sh` only formats the current commit's changes, so won't go as far as fixing entire files. So we'll have to use `clang-format` directly to fully fix these files.

Suggest a white-space only cleanup commit for all the mouse code:

`clang-format -i include/mouse.h src/ints/mouse*.cpp`

Here's the before-and-after of what I see (note the majority of formatting is all in the initial spacing, being moved to tabs):

![2022-10-13_08-06](https://user-images.githubusercontent.com/1557255/195634988-a99ea48b-0d15-4bd9-9368-f5990a19223a.png)

I know you've done a lot of careful alignment; most of it should still look pretty good after. The assignment aligner space-pads the left-hand-side by the longest consecutive type-and-variable name, so it produces this:

![2022-10-13_08-13](https://user-images.githubusercontent.com/1557255/195636892-fe2a9425-c397-41bf-9f2a-5e8ddeb54ee8.png)

When I see this, usually I'm using different types and just have to separate them with a new line (to group each chunk of variables by like-type):

![2022-10-13_08-17](https://user-images.githubusercontent.com/1557255/195637996-d92dd96e-845c-4cce-aec5-0d866b1cc541.png)

FeralChild64 replied:
Done - some code needed manual formatting update, nut not much.

FeralChild64 commented on src/hardware/mouse/mouse.cpp:
```diff
@@ -0,0 +1,673 @@
+/*
+ *  Copyright (C) 2022       The DOSBox Staging Team
+ *  Copyright (C) 2002-2021  The DOSBox Team
+ *
```
OK, added a README.md with some general description.

FeralChild64 commented on src/hardware/mouse/mouse.cpp:
```diff
@@ -0,0 +1,673 @@
+
+
+bool mouse_seamless_driver = false;
+bool mouse_seamless_user   = false;
```
OK, added helper access routines. Nevertheless, this shouldn't be exposed to GFX/SDL subsystem at all, the `sdlmain.cpp` has too many responsibilities. But again - this is not a topic for this PR.

FeralChild64 commented on src/hardware/mouse/mouse_common.cpp:
```diff
@@ -0,0 +1,204 @@
+    const auto diff_ticks = PIC_Ticks - pic_ticks_start;
+    const auto diff_ms    =
+            std::chrono::duration_cast<std::chrono::milliseconds>(diff_time).count();
+
```
OK, removed `std::chrono` checks. We don't need a `PIC_FullIndex()` here - such precision would be an overkill, `PIC_Ticks` is enough.

FeralChild64 commented on src/hardware/mouse/mouse_manymouse.cpp:
```diff
@@ -0,0 +1,476 @@
+
+    case MANYMOUSE_EVENT_RELMOTION:
+        // LOG_INFO("MANYMOUSE #%u RELMOTION axis %d, %d", event.device, event.item, event.value);
+        if (no_interface || critical_only || (event.item != 0 && event.item != 1))
```
I have added some more comments - 0 means 'x' axis, 1 means 'y'. At least on Linux (and at least my mouse) generates this event with other IDs for wheel scroll, in addition to `MANYMOUSE_EVENT_SCROLL`.

FeralChild64 commented on src/hardware/mouse/mouse_config.cpp:
```diff
@@ -0,0 +1,288 @@
+    // General configuration
+
+    prop_multi = secprop.AddMultiVal("mouse_sensitivity", only_at_start, ",");
+    prop_multi->Set_help("Default mouse sensitivity. Works exponentially, add 10 to double the effect.\n"
```
Done - added alternative (floating point) values support.

kcgen commented on src/hardware/mouse/mouse_config.cpp:
```diff
@@ -0,0 +1,349 @@
+		int32_t ret_val          = base_value; // default
+		bool invalid_input       = false;
+
+		if (str.find('.') != std::string::npos) {
+			// Parameter supplied in a form of floating point
+			float tmp_float = 1.0f;
+			try {
+				tmp_float = std::stof(str);
+			} catch (...) {
+				invalid_input = true;
+				tmp_float     = 1.0f;
+			}
+			// We need 'tmp_float' to be positive for further calculations
+			if (tmp_float < 0.0f) {
+				tmp_float = -tmp_float;
+				ret_val   = -1;
+			} else if (tmp_float > 0.0f)
+				ret_val = 1;
+			else
+				ret_val = 0;
+			// Now calculate user steps as a logarithm
+			if (ret_val != 0) {
+				tmp_float = std::log(tmp_float) / std::log(2.0f);
+				tmp_float *= mouse_predefined.sensitivity_double_steps;
+				tmp_float += static_cast<float>(base_value);
+				const auto tmp_rounded = std::lround(
+				        std::max(tmp_float, 1.0f));
+				ret_val = clamp_to_int8(ret_val * tmp_rounded);
+			}
+		} else if (!str.empty()) {
+			// Parameter supplied in a form of integer
+			try {
+				ret_val = std::stoi(str);
+			} catch (...) {
+				invalid_input = true;
+				ret_val       = base_value;
+			}
+		}
```
@FeralChild64 , I forgot that users of the original sensitivity will expect percentages instead of a floating-point scalar (so we should probably keep that interface).

To answer my previous feedback comment:
 - User wants to cut sensitivity into 1/3rd? Use 33
 - User wants to boost sensitivity by 50%? Use 150
 - User wants 8-fold sensitivity? Use 800
 - User wants to inverted behavior? Use -100 (for the Y-axis)

Hopefully it means we can just divide the user's int16_t value by 100.0f to get floating-point scale, and your existing code will apply straight-away.

Very sorry for this churn (but I think it makes sense) - as other conf settings (like volume) are also percentages (and not scalars).

FeralChild64 replied:
To keep it clear - you want me to change the current floating point code to accept percentage? Or to get rid of the exponential system completely? I wanted to have values 0-99 (exponential sensitivity) for:
- easier display in the command line tool (less characters needed)
- easier to create a GUI with a slider; as changing sensitivity from 100% to 150% will bring a very noticeable effect, but change from 500% to 550% won't

I don't know how the volume is implemented in DOSBox Staging, but usually it uses decibels, which are also logarithmic.

kcgen replied:
> you want me to change the current floating point code to accept percentage?

Yes. All DOSBoxes support percentage based mouse sensitivity (and mixer volumes):
 - "10" -> 1/10th of original
 - "50" -> 50% or half of original
 - "200" -> 2-fold more than original
 - "-100" -> invert the axis, at original sensitivity

A number in any form, provided without prefixed units, should be interpreted as this percentage.

For exponential input, I agree with all the benefits, and yes, all DOSBoxes also support decibel volume input when the number is prefixed with a `d`.
 - "d0" -> original/default level
 - "d-3" -> -3 dB is roughly half of original
 - "d3" -> +3 dB is roughly 2-fold more than original

I think a similar exponential prefixed-approach could be used for the mouse sensitivity.

> I wanted to have values 0-99 (exponential sensitivity)

The audio code parses the percent and decibel using the `parse_value` and `parse_prefixed_value` functions. Here's a simple adaptation if we assumed an `x` prefix for the mouse exponential input:

```
#include <string_utils.h>

const std::string arg = // user input;

// did the user provide a non-prefixed percentage?
if (auto v = parse_value(arg, -10000, 10000); v)
  exponential = percent_to_exponential(*val); 
 
// did the user provide an 'x' prefixed exponential?
else if (v = parse_prefixed_value('x', user_input, 0, 99); v)
  exponential = *v;
```

kcgen commented on src/hardware/mouse/mouse.cpp:
```diff
@@ -0,0 +1,652 @@
+			pattern_regex << ".";
+		else if (character == '*')
+			pattern_regex << ".*";
+		else if ((character >= '0' && character <= '9') ||
+		         (character >= 'a' && character <= 'z') ||
+		         (character >= 'A' && character <= 'Z'))
```
is alpha-numeric: `int isalnum( int ch );` from `<cctype>` maybe can do the job? 

https://en.cppreference.com/w/cpp/string/byte/isalnum

FeralChild64 replied:
Changed.

kcgen commented on src/hardware/mouse/mouse_common.cpp:
```diff
@@ -0,0 +1,184 @@
+	// one more small touch of 20th century PC computing history :)
+}
+
+uint8_t MOUSE_GetDelayFromRateHz(const uint16_t rate_hz)
+{
+	assert(rate_hz);
+	const auto tmp = std::lround(1000.0f / MOUSE_ClampRateHz(rate_hz));
+	return static_cast<uint8_t>(tmp);
```
if rate_hz is updates/s
-> substituting: 
(1000 ms / s) / (updates/s)

-> the seconds cancel out
ms / update

Maybe rename `tmp` to `period_in_ms`, to be more descriptive of the conversion?

FeralChild64 replied:
Changed.

kcgen commented on src/hardware/mouse/mouse_queue.cpp:
```diff
@@ -0,0 +1,314 @@
+	MouseQueue::GetInstance().Tick();
+}
+
+MouseQueue &MouseQueue::GetInstance()
+{
+	static MouseQueue *instance = nullptr;
+	if (!instance)
+		instance = new MouseQueue();
+	return *instance;
+}
```
Is the `instance` deleted at some point? (I couldn't find it, but might be in collapsed code).

FeralChild64 replied:
No, it isn't - this is a singleton. But I'll try to rework this to make the address sanitizer happy.

kcgen replied:
Maybe it doesn't need to be allocated on the heap?

```c++
static MouseQueue instance();
return instance;
```

FeralChild64 replied:
Changed obejct instance to static - also for `ManyMouseGlue` and concrete mouse interface classes.

kcgen commented:
Just minor comments left; also couldn't tell where these `new`d objects get `delete`d:

![2022-10-14_22-30](https://user-images.githubusercontent.com/1557255/195970608-71792cc7-ff20-471d-ad82-4da8793a4e85.png)

If there's a mouse shutdown or destructor somewhere, could be used to check and `delete` them.  Or we could use `std::unique_ptr` and have them cleanup themselves.

kcgen commented:
Just minor comments left :rocket:

kcgen commented:
Building with ASAN reports the leaks:

![2022-10-14_22-36](https://user-images.githubusercontent.com/1557255/195970795-c922606a-2fe1-4371-b0ee-7d4286db525b.png)

```shell
meson setup -Dbuildtype=debug -Db_sanitize=address -Doptimization=0 -Db_lundef=false -Dc_args=-fsanitize-recover=all -Dcpp_args=-fsanitize-recover=all -Dunit_tests=disabled build/clang-addresses

meson compile -C build/clang-addresses
```

I then did a `mousectl dos -map`, and then exited.

FeralChild64 commented:
@kcgen These ASAN reports were due to few objects allocated during startup, but not freed - but they were used till the end of the lifecycle. Changed objects to be static. BTW, I had to use release build, with debug build I'm getting an assert et the end:

```
dosbox: ../../src/gui/sdlmain.cpp:700: void log_display_properties(int, int, const std::optional<SDL_Rect> &, const SCALING_MODE, const SCREEN_TYPES, const PPScale &): Assertion `target_w > 0 && target_h > 0' failed.
```

Relevant part of my configuration file:
```ini
[sdl]
fullscreen          = true
display             = 0
fullresolution      = desktop
windowresolution    = 640x400
viewport_resolution = fit
window_position     = auto
window_decorations  = true
transparency        = 0
host_rate           = auto
vsync               = false
vsync_skip          = 7000
presentation_mode   = auto
output              = opengl
texture_renderer    = auto
capture_mouse       = seamless middlerelease
waitonerror         = true
priority            = auto auto
mute_when_inactive  = false
pause_when_inactive = false
mapperfile          = mapper-sdl2-0.80.0-alpha.map
screensaver         = auto
```

kcgen commented on src/hardware/mouse/mouse_manymouse.cpp:
```diff
@@ -0,0 +1,481 @@
+		HandleEvent(event);
+
+	// Report accumulated mouse movements
+	for (uint8_t idx = 0; idx < rel_x.size(); idx++) {
```
Suggest using a `size_t` for idx, or asserting that `rel_x.size()` is no bigger than `UINT8_MAX`.

FeralChild64 replied:
Added assertion.

kcgen commented on src/hardware/mouse/mouse_manymouse.cpp:
```diff
@@ -0,0 +1,481 @@
+		std::string name;
+		UTF8_RenderForDos(name_utf8, name);
+
+		const char character_nbsp  = 0x7f; // non-breaking space
+		const char character_space = 0x20;
+
+		for (auto pos = name.size(); pos > 0; pos--) {
+			// Replace non-breaking space with a regular space
+			if (name[pos - 1] == character_nbsp)
+				name[pos - 1] = character_space;
+			// Remove non-ASCII and control characters
+			if (name[pos - 1] < character_space ||
+			    name[pos - 1] >= character_nbsp)
+				name.erase(pos - 1, 1);
+		}
+
+		// Try to rework into something useful names in the forms
+		// 'FooBar Corp FooBar Corp Incredible Mouse'
+		size_t pos = name.size() / 2 + 1;
+		while (--pos > 2) {
+			if (name[pos - 1] != ' ')
+				continue;
+			const std::string tmp = name.substr(0, pos);
+			if (name.rfind(tmp + tmp, 0) == std::string::npos)
+				continue;
+			name = name.substr(pos);
+			break;
+		}
+
+		// ManyMouse should limit device names to 64 characters,
+		// but make sure name is indeed limited in length, and
+		// strip trailing spaces
+		name.resize(std::min(static_cast<size_t>(64), name.size()));
+		while (!name.empty() && name.back() == ' ')
+			name.pop_back();
+
```
This string processing is pretty involved and not too easy to follow (position division and addition, then substr'ing on position, etc).

Maybe just some basic cleanup - like `trim(name)` and `strip_punctuation(name)` is sufficient for the user to understand their mouse name?

```

void trim(std::string &str, const char trim_chars[] = " \r\t\f\n");
void strip_punctuation(std::string &str);

```

in `string_utils.h`.

kcgen replied:
(or if your routine does genel string sanitization, consider moving it to `string_utils.h/cpp`.)

FeralChild64 replied:
The `trim` was useful. As for punctuations, we want to keep them. Used `std::replace` to simplify the code a bit, improved comments. Unfortunately, not everything could be simplified.

The name cleanup is specific to mouse names, unfortunately.

kcgen commented on src/hardware/mouse/mouse_manymouse.cpp:
```diff
@@ -0,0 +1,481 @@
+{
+	// Wait a little to speedup screen update
+	const auto pic_ticks_start = PIC_Ticks;
+	while (PIC_Ticks >= pic_ticks_start && PIC_Ticks - pic_ticks_start < 50)
```
What's 50 represent? (move to constexpr)

FeralChild64 replied:
This is a time to wait idle in PIC ticks. Moved to `constexpr` and added a comment.

kcgen commented on src/hardware/mouse/mouse_manymouse.cpp:
```diff
@@ -0,0 +1,481 @@
+		while (!name.empty() && name.back() == ' ')
+			name.pop_back();
+
+		physical_devices.emplace_back(MousePhysical(name));
```
emplace_back calls its contained type's constructor with the given value.

`physical_devices` holds a type `MousePhysical`, so emplace_back is calling:

`MousePhysical(MousePhysical(name))`

The inner constructor creates a temporary, which is then passed into `MousePhysical` assignment/copy constructor. 

So you can do `.emplace_back(name)`.

FeralChild64 replied:
OK, changed. Couldn't do the same with `MousePhysicalInfoEntry`, unfortunately, due to private constructor.

kcgen commented on src/hardware/mouse/mouse_config.cpp:
```diff
@@ -0,0 +1,350 @@
+#endif
+};
+
+static const std::vector<std::string> list_models_com = {
+        "2button",
+        "3button",
+        "wheel",
+        "msm",
+        "2button+msm",
+        "3button+msm",
+        "wheel+msm",
+};
```
Suggest lifting these strings to their own constexprs:

```

auto model_2_button_name = "2button";

auto model_3_button_name = "3button";

auto model_wheel_name = "wheel";

...

```

Then these can be used to construct the vector's content and also do direct string-to-char[] comparisons in the all the if tests below (instead of using harcoded indexes).

FeralChild64 replied:
Did something like that, also for the PS/2 mouse models. Additionally, de-duplicated some code.

kcgen commented on src/hardware/mouse/mouse_queue.cpp:
```diff
@@ -0,0 +1,317 @@
+	} else if (!timer_in_progress) {
+		DEBUG_QUEUE("ActivateIRQ, in %s", __FUNCTION__);
+		// If no timer in progress, handle the event now
+		PIC_ActivateIRQ(12);
```
The magic `12` is used on 148 and 287 -> can you lift it to a local global `constexpr` (Is this the PS/2 mouse IRQ?)

FeralChild64 replied:
OK, added a constant

kcgen reviewed (changes requested):
Ok.. phew - finally completed a detailed review.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,543 @@
+        return false;
+    }
+
+    uint16_t value_hz = static_cast<uint16_t>(tmp);
```
Changed.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,543 @@
+	params.erase(params.begin()); // pop
+
+    // Check that all interfaces are unique
+    std::set<MouseInterfaceId> tmp(list_ids.begin(), list_ids.end());
```
Changed.

FeralChild64 commented on src/hardware/mouse/mouseif_dos_driver.cpp:
```diff
@@ -0,0 +1,1727 @@
+	auto call_int33 = CALLBACK_Allocate();
+	// RealPt int33_location = RealMake(CB_SEG + 1,(call_int33 * CB_SIZE) -
+	// 0x10);
+	RealPt int33_location = RealMake(static_cast<uint16_t>(DOS_GetMemory(0x1) - 1),
```
Changed, also in PS/2 code.

FeralChild64 commented on src/hardware/mouse/mouseif_dos_driver.cpp:
```diff
@@ -0,0 +1,1727 @@
+	uint16_t raxpt = real_readw(SegValue(ss), static_cast<uint16_t>(reg_sp + 0x0a));
+	uint16_t rbxpt = real_readw(SegValue(ss), static_cast<uint16_t>(reg_sp + 0x08));
+	uint16_t rcxpt = real_readw(SegValue(ss), static_cast<uint16_t>(reg_sp + 0x06));
+	uint16_t rdxpt = real_readw(SegValue(ss), static_cast<uint16_t>(reg_sp + 0x04));
```
Changed.

FeralChild64 commented on src/hardware/mouse/mouseif_dos_driver.cpp:
```diff
@@ -0,0 +1,1727 @@
+
+	// Apply mouse movement to mimic host OS
+	float x = calculate(x_abs, mouse_video.res_x, mouse_video.clip_x);
+	float y = calculate(y_abs, mouse_video.res_y, mouse_video.clip_y);
```
Changed.

FeralChild64 commented on src/hardware/mouse/mouseif_dos_driver.cpp:
```diff
@@ -0,0 +1,1727 @@
+		                   state.background.pos_x) *
+		                          2);
+		address /= 2;
+		uint16_t cr = real_readw(BIOSMEM_SEG, BIOSMEM_CRTC_ADDRESS);
```
Changed.

FeralChild64 commented on src/hardware/mouse/mouseif_dos_driver.cpp:
```diff
@@ -0,0 +1,1727 @@
+{
+	state.inhibit_draw = false;
+	// Get the correct resolution from the current video mode
+	uint8_t mode = mem_readb(BIOS_VIDEO_MODE);
```
Changed.

FeralChild64 commented on src/hardware/mouse/mouseif_ps2_bios.cpp:
```diff
@@ -0,0 +1,525 @@
+	};
+
+	static const std::vector<uint8_t> seq_im = {200, 100, 80};
+	static const std::vector<uint8_t> seq_xp = {200, 200, 80};
```
These are just magic numbers, for unlocking the extended protocol. The unit doesn't matter here, so I would prefer not to add `_hz` suffix. But I have changed the names to be more descriptive.

kcgen replied:
Thanks! I was just visually pattern-matching, thinking: _"these numbers match some of the  mouse sampling rates, so I guess they're rates"_.

FeralChild64 commented on src/hardware/mouse/mouseif_ps2_bios.cpp:
```diff
@@ -0,0 +1,525 @@
+	if (new_rate_hz != 10 && new_rate_hz != 20 && new_rate_hz != 40 &&
+	    new_rate_hz != 60 && new_rate_hz != 80 && new_rate_hz != 100 &&
+	    new_rate_hz != 200) {
+		// Invalid parameter, set default
```
We can't reuse that list, it's different. And when InPort/Bus mouse emulation is introduced, it will have another distinct list. But I have put the allowed values into `std::vector`.

FeralChild64 commented on src/hardware/mouse/mouseif_ps2_bios.cpp:
```diff
@@ -0,0 +1,525 @@
+	case 4: cmd_set_sample_rate(80); break;
+	case 5: cmd_set_sample_rate(100); break;
+	case 6: cmd_set_sample_rate(200); break;
+	default: return false;
```
Changed, together with other rate list remark. Did the same change for mouse resolutions.

kcgen commented on src/hardware/mouse/mouse_interfaces.cpp:
```diff
@@ -0,0 +1,839 @@
+        : MouseInterface(MouseInterfaceId::PS2, mouse_predefined.sensitivity_ps2)
+{}
+
+InterfacePS2::~InterfacePS2() {}
```
(these `= default;`s can be moved into the header class definitions)

FeralChild64 commented on src/hardware/mouse/mouseif_ps2_bios.cpp:
```diff
@@ -0,0 +1,525 @@
+	case -2: return -1;
+	case -1: return -1;
+	case 1: return 1;
+	case 2: return 1;
```
Changed.

FeralChild64 commented on src/hardware/mouse/mouseif_ps2_bios.cpp:
```diff
@@ -0,0 +1,525 @@
+	case -4: return -6;
+	case -3: return -3;
+	case -2: return -1;
+	case -1: return -1;
```
Changed.

FeralChild64 commented on src/hardware/mouse/mouse_interfaces.cpp:
```diff
@@ -0,0 +1,839 @@
+                         mouse_predefined.sensitivity_com)
+{}
+
+InterfaceCOM::~InterfaceCOM() {}
```
Changed.

FeralChild64 commented on src/hardware/mouse/mouse_interfaces.cpp:
```diff
@@ -0,0 +1,839 @@
+        : MouseInterface(MouseInterfaceId::DOS, mouse_predefined.sensitivity_dos)
+{}
+
+InterfaceDos::~InterfaceDos() {}
```
Changed.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,543 @@
+
+bool MOUSECTL::CmdMap()
+{
+    assert(list_ids.size() >= 1);
```
Changed.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,543 @@
+            return CmdMap(list_ids[0], params[1]);
+
+    // CmdMap - interactive
+    if (list_ids.size() >= 1 && params.size() == 1)
```
Changed.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,543 @@
+    // CmdShow
+    if (list_ids.size() == 0 && params.size() == 0)
+        return CmdShow(false);
+    if (list_ids.size() == 0 && params.size() == 1)
```
Changed.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,543 @@
+        if (param_equal(0, "-reset"))
+            return CmdReset();
+        if (param_equal(0, "-s"))
+            return CmdSensitivity();
+        if (param_equal(0, "-sx"))
+            return CmdSensitivityX();
+        if (param_equal(0, "-sy"))
+            return CmdSensitivityY();
+        if (param_equal(0, "-r"))
+            return CmdMinRate();
```
As discussed on Discord, changed sensitivity concept to only accept values from -999 to +999, as percentage.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,543 @@
+        if (param_equal(0, "-sx"))
+            return CmdSensitivityX(static_cast<int8_t>(value));
+        if (param_equal(0, "-sy"))
+            return CmdSensitivityY(static_cast<int8_t>(value));
```
As discussed on Discord, changed sensitivity concept to only accept values from -999 to +999, as percentage.

FeralChild64 commented on src/hardware/mouse/mouse_interfaces.cpp:
```diff
@@ -0,0 +1,839 @@
+			power   = static_cast<float>(user_val - 50);
+			scaling = sensitivity_predefined;
+		} else if (user_val < 0) {
+			power   = static_cast<float>(-user_val - 50);
```
As discussed on Discord, changed sensitivity concept to only accept values from -999 to +999, as percentage.
This results in this piece of code being removed.

FeralChild64 commented on src/hardware/mouse/mouseif_ps2_bios.cpp:
```diff
@@ -0,0 +1,525 @@
+		CPU_Push16(packet[2]);
+		CPU_Push16(packet[3]);
+	}
+	CPU_Push16((uint16_t)0);
```
OK, changed.

FeralChild64 commented on src/hardware/mouse/mouse_interfaces.cpp:
```diff
@@ -0,0 +1,839 @@
+
+void MouseInterface::NotifyInterfaceRate(const uint16_t value_hz)
+{
+	interface_rate_hz = value_hz;
```
OK, changed.

FeralChild64 commented on src/hardware/mouse/mouse_interfaces.h:
```diff
@@ -0,0 +1,228 @@
+private:
+	MouseInterface()                                  = delete;
+	MouseInterface(const MouseInterface &)            = delete;
+	MouseInterface &operator=(const MouseInterface &) = delete;
```
OK, changed.

FeralChild64 commented on src/hardware/mouse/mouseif_ps2_bios.cpp:
```diff
@@ -0,0 +1,525 @@
+	if (dy < 0) {
+		dy          = static_cast<int16_t>(dy + 0x100);
+		mdat.sign_y = 1;
+	}
```
They are actually related to UINT8_MAX - so used these definitions.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,543 @@
+        if (value < -99 || value > 99) {
+            WriteOut(MSG_Get("SHELL_CMD_MOUSECTL_SYNTAX_SENSITIVITY"));
+            return false;
+        }
```
Changed to `std::stoi` + exception catching.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,543 @@
+    default:
+        assert(false); // missing implementation
+        return nullptr;
+    }
```
Turned into static. Also, I have moved this to `MouseControlAPI`, as it might be useful for the future GUI.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,543 @@
+    default:
+        assert(false); // missing implementation
+        return nullptr;
+    }
```
OK, changed.

FeralChild64 commented on src/hardware/mouse/mouse_manymouse.cpp:
```diff
@@ -0,0 +1,481 @@
+
+	ClearPhysicalMice();
+
+	for (uint8_t idx = 0; idx < num_mice; idx++) {
```
`num_mice` is initialized in `ManyMouseGlue::InitIfNeeded()`, which has a protection - so it shouldn't happen. For extra safety I have reworked the code so that `num_mice` is `uint8_t` now.

FeralChild64 commented on src/hardware/mouse/mouseif_ps2_bios.cpp:
```diff
@@ -0,0 +1,525 @@
+		MOUSE_NotifyStateChanged();
+		return true;
+	}
+}
```
OK, changed.

FeralChild64 commented on src/hardware/mouse/mouseif_dos_driver.cpp:
```diff
@@ -0,0 +1,1727 @@
+		return static_cast<int16_t>(x - 0x10000);
+	else
+		return static_cast<int16_t>(x);
+}
```
OK, changed. I think I did it to work around some compiler warnings - strangely, I can't see any now.

kcgen commented:
![Screenshot from 2022-10-21 13-14-35](https://user-images.githubusercontent.com/1557255/197281260-56d6033c-05dc-427c-8605-72a35b9cf596.png)

At the input prompt, I couldn't quit using Ctrl+F9.

I'm not sure if this input loop is in ManyMouse, but you can check for `shutdown_requested` going true, this is a system-wide `bool` that goes true on any exit action (GUI-close/keyboard hotkey/signal).

It's defined in `dosbox.h`, so should be available in most DOSBox-side code:

```c++
// The shutdown_requested bool is a conditional break in the parse-loop and
// machine-loop. Set it to true to gracefully quit in expected circumstances.
extern bool shutdown_requested;
```

kcgen commented on src/hardware/mouse/mouse_interfaces.cpp:
```diff
@@ -0,0 +1,818 @@
+
+const MousePhysical &MouseInterfaceInfoEntry::MappedPhysical() const
+{
+	const auto idx = Interface().GetMappedDeviceIdx();
```
`idx` is also a member of this `MouseInterfaceInfoEntry` class.
Should we update it, or is this a temporary?
(If it's a temporary, suggest a different name to avoid shadowing)

FeralChild64 replied:
No, this is just a temporary variable. Changed both names to avoid confusion.

kcgen commented on src/hardware/mouse/mouse_interfaces.cpp:
```diff
@@ -0,0 +1,818 @@
+	return rate_hz;
+}
+
+void MouseInterface::NotifyInterfaceRate(const uint16_t rate_hz)
```
`rate_hz` is also a member of this `MouseInterface` class. 
Suggest a new argument name to avoid ambiguity/shadowing in the `interface_rate_hz = rate_hz` statement.

FeralChild64 replied:
OK, changed argument name.

kcgen commented on src/hardware/mouse/mouse_queue.cpp:
```diff
@@ -0,0 +1,317 @@
+	if (!pic_ticks_start)
+		elapsed = 1;
+
+	auto calc_new_delay = [](const uint8_t delay, const uint8_t elapsed) {
+		return static_cast<uint8_t>((delay > elapsed) ? (delay - elapsed)
+		                                              : 0);
+	};
```
`delay` is also a member of the `MouseQueue`; suggest a new local name to avoid ambiguity/shadowing.

FeralChild64 replied:
OK, changed argument name.

kcgen commented on src/hardware/serialport/serialmouse.cpp:
```diff
@@ -28,452 +28,551 @@
+
+void CSerialMouse::BoostRate(const uint16_t rate_hz)
+{
+	if (!rate_hz || model == MouseModelCOM::NoMouse) {
+		rate_coeff = 1.0f;
+		return;
+	}
```
`model` user here is coming from the `CSerialMouse`'s member variable.

kcgen commented on src/hardware/serialport/serialmouse.cpp:
```diff
@@ -28,452 +28,551 @@
+		// In addition to byte_len, the mouse has to send
+		// 3 more bits per each byte: start, parity, stop
+
+		if (model == MouseModelCOM::Microsoft ||
+		    model == MouseModelCOM::Logitech || model == MouseModelCOM::Wheel)
+			// Microsoft-style protocol
+			// single movement needs exactly 3 bytes to be reported
+			return bauds / (static_cast<float>(byte_len + 3) * 3.0f);
+		else if (model == MouseModelCOM::MouseSystems)
```
But `model` here is an argument passed into the `estimate(...)` lamba.

FeralChild64 replied:
Changed, captured `this`.

kcgen commented on src/hardware/serialport/serialmouse.cpp:
```diff
@@ -28,452 +28,551 @@
+	};
+
+	// Calculate coefficient to match requested rate
+	rate_coeff = estimate(1200, port_byte_len, model) / rate_hz;
```
And then here `model` is from the `CSerialMouse`'s member (which becomes a copy inside estimate).

Suggest letting the lamba access the member variables by capturing `this` (I'm not sure why the c++ committee didn't make this the default, given the lamda function is already inside scope of the class just like the code around it :man_facepalming: ):

```c++
	auto estimate = [this](const uint16_t bauds) {
        // lamba now has access to port_byte_len and model
        ...
    };
    ...

    rate_coeff = estimate(1200) / rate_hz;

```

FeralChild64 replied:
OK, changed.

kcgen commented on src/dos/program_mousectl.h:
```diff
@@ -0,0 +1,76 @@
+/*
```
This one might have slipped the clang reformatting pass (using a lot of space-indenting) Here's what I get when reformat:

![2022-10-21_14-15](https://user-images.githubusercontent.com/1557255/197292981-6819c601-eb11-4eef-9e3c-64517dc42245.png)

FeralChild64 replied:
OK, reformatted.

kcgen commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,593 @@
+/*
```
This one might have slipped through the clang reformatting cleanup; still using a lot of space-indenting (here's what I get):

![2022-10-21_14-15_1](https://user-images.githubusercontent.com/1557255/197293015-c904901c-7cf6-4a2d-80de-1b2db01d0e66.png)

FeralChild64 replied:
OK, reformatted.

kcgen reviewed (changes requested):
Passing testing and the UBSAN build is also running clean!  :checkered_flag: 

Just some minor comments left.

kcgen commented on include/math_utils.h:
```diff
@@ -127,6 +127,33 @@ constexpr T1 left_shift_signed(T1 value, T2 amount)
 	return static_cast<T1>(shifted);
 }
 
+template <typename T>
+int8_t clamp_to_int8(const T val)
+{
+	const auto tmp = std::clamp(val,
+	                            static_cast<T>(INT8_MIN),
+	                            static_cast<T>(INT8_MAX));
+	return static_cast<int8_t>(tmp);
+}
+
+template <typename T>
+int16_t clamp_to_int16(const T val)
+{
+	const auto tmp = std::clamp(val,
+	                            static_cast<T>(INT16_MIN),
+	                            static_cast<T>(INT16_MAX));
+	return static_cast<int16_t>(tmp);
+}
+
+template <typename T>
+int32_t clamp_to_int32(const T val)
+{
+	const auto tmp = std::clamp(val,
+	                            static_cast<T>(INT32_MIN),
+	                            static_cast<T>(INT32_MAX));
+	return static_cast<int32_t>(tmp);
+}
+
```
I've added unit tests for these (on a branch), if you can cherry-pick them in: 2434cc89d32

kcgen replied:
And unsigned handling to `clamp_to` in 66091cfd7f0f90259552c639dbfda912e69be3de

FeralChild64 replied:
OK, cherry-picked - not closing the conversation, as I need to review the code if these templates can be reused elsewhere in the mouse code now.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,593 @@
+	}
+
+	// Maximum allowed user sensitivity value
+	const int16_t sensitivity_user_max = 999;
```
OK, changed.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,593 @@
+	MSG_Add("SHELL_CMD_MOUSECTL_SYNTAX_PATTERN",
+	        "Wrong syntax, only ASCII characters allowed in pattern.");
+	MSG_Add("SHELL_CMD_MOUSECTL_SYNTAX_SENSITIVITY",
+	        "Wrong syntax, sensitivity needs to be in -99 to +99 range.");
```
OK, fixed.

FeralChild64 commented on src/dos/program_mousectl.cpp:
```diff
@@ -0,0 +1,593 @@
+    }
+
+    const auto value_hz = static_cast<uint16_t>(tmp);
+    if (std::find(valid_list.begin(), valid_list.end(), value_hz) == valid_list.end()) {
```
OK, changed.

FeralChild64 commented on src/hardware/mouse/mouse.cpp:
```diff
@@ -0,0 +1,667 @@
+{
+	const auto &valid_list = GetValidMinRateList();
+	if (std::find(valid_list.begin(), valid_list.end(), value_hz) ==
+	    valid_list.end())
```
OK, changed.

FeralChild64 commented:
> At the input prompt, I couldn't quit using Ctrl+F9.
> 
> I'm not sure if this input loop is in ManyMouse, but if it's possible to check for `shutdown_requested`, this is a system-wide bool that goes `true` on any exit action (GUI-close/keyboard hotkey/signal).

@kcgen: Fixed, added `shutdown_requested` as a loop condition inside `ManyMouseGlue::ProbeForMapping()`

kcgen commented on src/gui/sdlmain.cpp:
```diff
@@ -4311,20 +4314,10 @@ void Config_Add_SDL() {
-	        "acceleration and sensitivity settings. This works in\n"
-	        "fullscreen or when the mouse is captured in window mode.");
+	Pmulti = sdl_sec->AddMultiVal("sensitivity", deprecated, ",");
+	Pmulti->Set_help("Moved [mouse] section; see documentation as values have different meaning now.");
```
-> "Moved **to** [mouse] section."

And (I think?) we can drop the "see documentation as values have different meaning now." portion?

FeralChild64 replied:
Right, changed.

kcgen reviewed (approved):
Looks good! Let's get this merged 🚀

Kappa971 commented:
Have the messages of the new `mousectl` command been made translatable?

FeralChild64 commented:
@Kappa971: I'm not sure if I understand correctly. The `config -wl outfile.txt` correctly writes out the message file for me.

Kappa971 commented:
> @Kappa971: I'm not sure if I understand correctly. The `config -wl outfile.txt` correctly writes out the message file for me.

I haven't downloaded the new build yet but I remember that with the mixer command, it was necessary to create a separate PR to translate all the new strings in `mixer` and `mixer /listmidi` while the configuration file and `mixer /?` it was already ok.
In this case surely `mousectl /?` and the config file are updated, but the strings of the screenshot?

Torinde commented:
@nemo93 Besides Settlers - also Serious Sam (Win9x game) [supports dual mouse](https://ofp-faguss.com/blog/serious-sam-configuration/).

Kappa971 commented:
> @nemo93 Besides Settlers - also Serious Sam (Win9x game) [supports dual mouse](https://ofp-faguss.com/blog/serious-sam-configuration/).

Serious Sam came out in 2001, it's not a DOS game.
